### DISCUSSION TITLE
Any guide/blog post/resources where I can see how to setup Match to play along with Amazon S3 Bucket? Or am I missing something?

### DISCUSSION THREAD

**Discussion by piscespieces** [fastlane/fastlane#20485] [Q&A]
Currently trying to store my certs and provisioning profiles in a S3 Bucket, but when I run the fastlane match development/appstore I get the following log

```
 Error accessing file, this might be due to fastlane's directory handling
[16:31:19]: Check out https://docs.fastlane.tools/advanced/#directory-behavior for more details

/opt/homebrew/Cellar/fastlane/2.208.0/libexec/gems/aws-sdk-core-3.131.3/lib/seahorse/client/plugins/response_target.rb:67:in `initialize': No such file or directory @ rb_sysopen - /var/folders/td/2ks8dv9s3fl42qzrdmlg9r0w0000gn/T/d20220719-23830-ooiwmv/cert-test/ (Errno::ENOENT)
```

This is how my env looks like

```
MATCH_PASSWORD=1234
MATCH_KEYCHAIN_PASSWORD=1234
MATCH_S3_BUCKET=mobile-application-certs
MATCH_S3_REGION=us-east-1
MATCH_S3_ACCESS_KEY=xxx
MATCH_S3_SECRET_ACCESS_KEY=xxx
```

And this is how my Matchfile looks like

```
# git_url("https://github.com/springbig/springbig-mobile.git")
# git_branch("master")
# storage_mode("git")
storage_mode("s3")
s3_bucket(ENV["MATCH_S3_BUCKET"])
s3_region(ENV["MATCH_S3_REGION"])
s3_access_key(ENV["MATCH_S3_ACCESS_KEY"])
s3_secret_access_key(ENV["MATCH_S3_SECRET_ACCESS_KEY"])

type("appstore")
username("test@test.com")
app_identifier("org.example.asd.com")
```

I haven't been able to find any good resource on how to do this.

**Comment by whs** [comment]
I recently fixed this issue by removing the folder (non-recursively) in S3.

In S3 there's no concept of a folder, just files with `/` in their name. But in GUI tools you'd need to create and enter a folder to place files inside, so when you create a directory in those tools it's actually creating an empty file with trailing slash in its name.

You can confirm the presence of such marker file by using `aws s3 ls s3://bucketname/prefix/` and you'll see file of size 0 with no name. Or you can just  use `aws s3 head-bucket --bucket bucketname --key prefix/` which will throw an error if the marker file does not exists.

Now, when you use `match`, fastlane download the entire S3 prefix onto the disk. This include those files with trailing slash, but you can't have a file with slash in its name so the download fails.

I believe the best way is to fix fastlane from trying to download these files with trailing slash (adding short circuit somewhere around [here](https://github.com/fastlane/fastlane/blob/e14064258cd784d65e30c19fe137ecbe5f5c8da7/match/lib/match/storage/s3_storage.rb#L105)). Alternatively, a workaround is to delete those files. In OP's case, it's probably

```sh
aws s3 rm s3://bucketname/d20220719-23830-ooiwmv/cert-test/
```

(Since we're not using `--recursive`, files inside this 'folder' are unaffected)

**Reply by whs** [reply]
Also offtopic, but should be useful to someone looking to use AWS AssumeRoleWithWebIdentity with Fastlane: Just set the S3 access key/secret to `nil` and set the relevant environment variables. Now you don't have to worry about leaking permanent AWS credentials!
